packages/devtools/src/types.ts

```diff
@@ -0,0 +1,18 @@
+export interface Account {
+  privateKey: string;
+  amount?: string;
+}
+
+export interface Deployment {
+  artifact: any;
+  arguments?: any[];
```

Comment:
Actually, making it mandatory to supply it might be more pain than benefit.

Reply:
I'm having trouble seeing this coupling.

> the code in `ganache-server.ts` isn't coupled to the constructor signatures of the contracts being deployed

AFAICT the code in `ganache-server.ts` is just detecting if the contract has any arguments, it doesn't care about what they are. So the calling code (in this case `deployer` in `nitro-protocol`) specifying these arguments as just the dependent contracts' names in a list, or their `DeployedArtifact` representations (i.e. said contracts' names _and_ their addresses) via a function doesn't seem to be a concern.

The function implementation would require access to a `deployedArtifacts` object holding the dependent contracts' addresses (as referenced in the second code snippet above), which is probably not something the _calling_ code should care about, right?

And to continue on that note:

> I'm not sure that's clear to someone without the context that the only constructor arguments are other contract addresses.

in the current circumstances of deploying the contracts, it just so happens to be other contracts' addresses, but the arguments can be anything, right? (hence, the usage of the `any` type)

Let me know if I got this backwards

Reply:
> I'm having trouble seeing this coupling.

Take [this line](https://github.com/statechannels/monorepo/pull/424/files/0c50e8a6165155e6a6cc07ce0d8db5f216e6542d#diff-9d77fecdba8b16354b02c5308eee735bR88).

It only works because the only arguments currently used in constructors are addresses of other contracts that are deployed. Here is [one](https://github.com/statechannels/monorepo/blob/master/packages/nitro-protocol/contracts/ERC20AssetHolder.sol#L19) of them.

Let's say you change that constructor to 
```
    constructor(uint userId, address _AdjudicatorAddress, address _TokenAddress)
```
At that point, [this line](https://github.com/statechannels/monorepo/blob/0c50e8a6165155e6a6cc07ce0d8db5f216e6542d/packages/devtools/src/utils/ganache-server.ts#L88) no longer works: `arg` in that case some arbitrary number, and is not in `deployedArtifacts` object. So, [this check](https://github.com/statechannels/monorepo/blob/0c50e8a6165155e6a6cc07ce0d8db5f216e6542d/packages/devtools/src/utils/ganache-server.ts#L87) fails, and the `userId` that I supplied as an argument to the deployer doesn't get pushed onto the argument list that's passed to the constructor.

Essentially, the existing deployment code in `ganache-server.ts`
- assumes that constructor arguments are only addresses of contracts that have already been deployed
- doesn't actually accept the arguments themselves, but instead accepts the _names_ of the contract, which it converts to an address by looking at `deployedArtifacts`

> The function implementation would require access to a `deployedArtifacts` object holding the dependent contracts' addresses (as referenced in the second code snippet above), which is probably not something the _calling_ code should care about, right? Seems like the only thing the calling code should care about is specifying which other contracts it depends on.

The functional implementation doesn't require access to the `deployedArtifacts` object itself; it's an instruction on how to compute the arguments, given the `deployedArtifacts`. The reality is that the constructor for `EthAssetHolder` requires the address of the artifacts that have been already deployed, ie. it requires knowledge about `deployedArtifacts`. 


> in the current circumstances of deploying the contracts, it just so happens to be other contracts' addresses, but the arguments can be anything, right? (hence, the usage of the `any` type)

Yes, the arguments could be anything, but as pointed out above, if we change the constructor signature, arguments that are not contract addresses would not be passed on to the constructor.

Reply:
Yeah I had a sense of what you were referring to and I didn't like that coupling either. Updated it to be more decoupled as you suggested.